crates/namada/src/ledger/protocol/mod.rs

```diff
@@ -432,166 +482,168 @@ where
+            .map_err(Error::StorageError)?;
 
-            if let Some(post_bal) = balance.checked_sub(fees) {
-                token_transfer(
-                    state,
+            let (post_bal, valid_batched_tx_result) = if let Some(post_bal) =
+                balance.checked_sub(fees)
+            {
+                fee_token_transfer(
+                    shell_params.state,
                     &wrapper.fee.token,
                     &wrapper.fee_payer(),
                     block_proposer,
                     fees,
-                )
-                .map_err(|e| Error::FeeError(e.to_string()))?;
+                )?;
 
-                let target_post_balance = Some(
-                    namada_token::read_balance(
-                        state,
+                (Some(post_bal), None)
+            } else {
+                // See if the first inner transaction of the batch pays the fees
+                // with a masp unshield
+                if let Ok(Some(valid_batched_tx_result)) =
+                    try_masp_fee_payment(shell_params, tx, tx_index)
+                {
+                    // NOTE: Even if the unshielding was succesfull we could
+                    // still fail in the transfer (e.g. cause the unshielded
+                    // amount is not enough to cover the fees). In this case we
+                    // want do drop the changes applied by the masp transaction
+                    // and try to drain the fees from the transparent balance.
+                    // Because of this we must NOT propagate errors from within
+                    // this branch
+                    let balance = crate::token::read_balance(
+                        shell_params.state,
                         &wrapper.fee.token,
-                        block_proposer,
-                    )
-                    .map_err(Error::StorageError)?
-                    .into(),
-                );
+                        &wrapper.fee_payer(),
+                    );
 
-                state.write_log_mut().emit_event(
-                    TokenEvent {
-                        descriptor: FEE_PAYMENT_DESCRIPTOR,
-                        level: EventLevel::Tx,
-                        token: wrapper.fee.token.clone(),
-                        operation: TokenOperation::Transfer {
-                            amount: fees.into(),
-                            source: UserAccount::Internal(wrapper.fee_payer()),
-                            target: UserAccount::Internal(
-                                block_proposer.clone(),
-                            ),
-                            source_post_balance: post_bal.into(),
-                            target_post_balance,
-                        },
-                    }
-                    .with(HeightAttr(current_block_height))
-                    .with(TxHashAttr(wrapper_tx_hash)),
-                );
+                    // Ok to unwrap_or_default. In the default case, the only
```

Comment:
maybe its safer to panic. If there is a bug, this would allow it to be silent.

Reply:
I've added an unwrap to the previous operation (the storage read) in 9698bb7a0da969ea45b68de6fa601f5689cd7899